candidategenerator.js

```diff
@@ -28,14 +28,82 @@ var wNum = 0;
+	 - recursiveCategory (string)
+	 - terminalCategory (string)
    - addTones (string). Possible values include:
 	 		- "addJapaneseTones"
 			- "addIrishTones_Elfner"
 			- "addIrishTones_Kalivoda"
+	- noUnary (boolean): if true, don't create any nodes that immediately dominate only a single terminal.
+	- requireRecWrapper (boolean). Formerly "requirePhiStem"
 */
 window.GEN = function(sTree, words, options){
 	options = options || {}; // if options is undefined, set it to an empty object (so you can query its properties without crashing things)
 
+	options.recursiveCategory = options.recursiveCategory || "phi"; //sets the default of recursiveCategory option to "phi"
+
+	//set default root root category based on options passed
+	if(!options.rootCategory){ //root category is unspecifed
+		//recursiveCategory is specified as "w"
+		if(options.recursiveCategory === "w"){
+			options.rootCategory = "phi";
+		}
+		else{
+			options.rootCategory = "i";
+		}
+	}
+
+	//set default terminal category based on options passed
+	if(!options.terminalCategory){ //terminalCategory is unspecified
+		//terminalCategory unspecified, recursiveCategory specified as "i"
+		if(options.recursiveCategory === "i"){
+			options.terminalCategory = "phi";
+		}
+		//terminalCategory unspecified, recursiveCategory specified as "w"
+		else if(options.recursiveCategory === "w"){
+			options.terminalCategory = "Ft";
+		}
+		//neither terminalCategory nor recursiveCategory is specified, default to "w"
+		else{
+			options.terminalCategory = "w";
+		}
+	}
+
+	/* the prosodic hierarchy should include the categories specified in
+	 * options.rootCategory, options.recursiveCategory and options.terminalCategory
+	 */
+	if(options.rootCategory || options.recursiveCategory || options.terminalCategory){
+		//the specified root category should probably be highest
+		if(pCat.indexOf(options.rootCategory)<0)
+			pCat.unshift(options.rootCategory);
+		if(pCat.indexOf(options.recursiveCategory)<0 && options.rootCategory !== options.recursiveCategory){
+			//the specified recursive category should be after iota
+			try{
+				var afterI = pCat.indexOf("i") + 1;
+				pCat.splice(afterI, 0, options.recursiveCategory);
+			}
+			catch(err){
+				console.error("make sure pCat contains your recursive category.");
+			}
+		}
+		//the specified terminal category should be right above "syll"
+		if(pCat.indexOf(options.terminalCategory)<0 && options.terminalCategory !== options.recursiveCategory && options.rootCategory !== options.terminalCategory)
+			pCat.splice(-1, 0, options.terminalCategory);
+	}
+
+	if(options.rootCategory === options.recursiveCategory && options.obeysNonrecursivity){
+		alert("Warning:\nYou have instructed GEN to produce non-recursive trees and to produce trees where the root node and\nintermediate nodes are of the same category. Some of the trees GEN produces will be recursive.")
+	}
+
+	if(options.rootCategory === options.recursiveCategory && options.obeysNonrecursivity){
```

Comment:
Should this be (options.rootCategory === options.terminalCategory)?

Reply:
Also, in this case, *all* of the trees will be recursive.

Reply:
You are right, thanks for catching that